### ISSUE TITLE
Tensorflow type 21 not convertible to numpy dtype.

### ISSUE THREAD

**Issue by hlsfin** [ray-project/ray#18479]
Anyone getting 'Tensorflow type 21 not convertible to numpy dtype.'
when trying to tune their CNN classifier model on TF 2/keras ? 
Thank you.

**Comment by richardliaw** [comment]
Can you provide a simple reproducible script?

**Comment by hlsfin** [comment]
import argparse
import os

from filelock import FileLock
from tensorflow.keras.datasets import mnist

import ray
from ray import tune
from ray.tune.schedulers import AsyncHyperBandScheduler
from ray.tune.integration.keras import TuneReportCallback
from tensorflow.keras.layers import Input, Conv2D, BatchNormalization, ReLU
from tensorflow.keras.layers import MaxPool2D, Dropout, Dense, Flatten
from tensorflow.keras.layers.experimental.preprocessing import Rescaling
import tensorflow as tf

def split_dataset(dataset, train_split, val_split, test_split):
#     shuffle = True
#     shuffle_size = 10000
    assert(train_split + val_split + test_split) == 1

    dataset_size = len(dataset)

#     if shuffle:
#         dataset = dataset.shuffle(shuffle_size, seed=12)

    train_size = int(train_split * dataset_size)
    val_size = int(val_split * dataset_size)

    train_dataset = dataset.take(train_size)
    val_dataset = dataset.skip(train_size).take(val_size)
    test_dataset = dataset.skip(train_size).skip(val_size)

    return(train_dataset, val_dataset, test_dataset)
EPOCHS = 200
CHANNELS = 3
BATCH_SIZE = 32
#IMAGE_SIZE = 256
IMAGE_SIZE=224

dataset = tf.keras.preprocessing.image_dataset_from_directory(
    "dataset",
    seed = 123,
    shuffle=True,
    label_mode='categorical',
    image_size=(IMAGE_SIZE, IMAGE_SIZE),
    batch_size=BATCH_SIZE
    
)
train_split = 0.6
val_split = 0.4
test_split = 0.0

train_ds, validation_ds, test_ds = split_dataset(dataset, train_split, val_split, test_split)

def train_mnist(config):
  


    input_shape = (BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE, CHANNELS)
    total_class = 19

    model = tf.keras.models.Sequential.Sequential([
  

    tf.keras.layers.Conv2D(32, kernel_size = (3, 3), activation = 'relu', input_shape= input_shape),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, kernel_size = (3,3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, kernel_size = (3,3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, kernel_size = (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, kernel_size = (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, kernel_size = (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(64, activation = 'relu'), 
    tf.keras.layers.Dense(total_class, activation = 'softmax'),
    ])
 
    model.compile(
        loss="sparse_categorical_crossentropy",
        optimizer=tf.keras.optimizers.SGD(
            lr=config["lr"],
        metrics=["accuracy"]))

    model.fit(
        train_ds,
        batch_size=32,
        epochs=10,
        verbose=0,
        validation_data=(validation_ds),
        callbacks=[TuneReportCallback({
            "mean_accuracy": "accuracy"
        })])



ray.init(num_cpus=4)
# sched = AsyncHyperBandScheduler(
#     time_attr="training_iteration", max_t=400, grace_period=20)

analysis = tune.run(
    train_mnist,
    name="exp",
    
    metric="mean_accuracy",
    mode="max",
    stop={
        
        "training_iteration": 3
    },
 
    config={
        
        "lr": tune.uniform(0.001, 0.1)
    
    }


    )
print("Best hyperparameters found were: ", analysis.best_config)

**Comment by hlsfin** [comment]
import ray
import tensorflow as tf
from ray import tune
from ray.tune.integration.keras import TuneReportCallback


def split_dataset(dataset, train_split, val_split, test_split):
    assert (train_split + val_split + test_split) == 1

    dataset_size = len(dataset)

    train_size = int(train_split * dataset_size)
    val_size = int(val_split * dataset_size)

    train_dataset = dataset.take(train_size)
    val_dataset = dataset.skip(train_size).take(val_size)
    test_dataset = dataset.skip(train_size).skip(val_size)

    return (train_dataset, val_dataset, test_dataset)


EPOCHS = 200
CHANNELS = 3
BATCH_SIZE = 32

IMAGE_SIZE = 224

dataset = tf.keras.preprocessing.image_dataset_from_directory(
    "dataset",
    seed=123,
    shuffle=True,
    label_mode='categorical',
    image_size=(IMAGE_SIZE, IMAGE_SIZE),
    batch_size=BATCH_SIZE

)
train_split = 0.6
val_split = 0.4
test_split = 0.0

train_ds, validation_ds, test_ds = split_dataset(dataset, train_split, val_split, test_split)


def train_mnist(config):
    input_shape = (BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE, CHANNELS)
    total_class = 19

    model = tf.keras.models.Sequential.Sequential([

        tf.keras.layers.Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=input_shape),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(64, activation='relu'),
        tf.keras.layers.Dense(total_class, activation='softmax'),
    ])

    model.compile(
        loss="sparse_categorical_crossentropy",
        optimizer=tf.keras.optimizers.SGD(
            lr=config["lr"],
            metrics=["accuracy"]))

    model.fit(
        train_ds,
        batch_size=32,
        epochs=10,
        verbose=0,
        validation_data=(validation_ds),
        callbacks=[TuneReportCallback({
            "mean_accuracy": "accuracy"
        })])


ray.init(num_cpus=4)

analysis = tune.run(
    train_mnist,
    name="exp",

    metric="mean_accuracy",
    mode="max",
    stop={
        "training_iteration": 3},
    config={

        "lr": tune.uniform(0.001, 0.1)

    }

)
print("Best hyperparameters found were: ", analysis.best_config)

**Comment by xwjiang2010** [comment]
Hi @hlsfin,
Thanks for the script. Please take a look [here](https://discuss.ray.io/t/using-raytune-with-tensorflow-recommenders-library-and-pbt-internalerror/299).
I believe Dataset object is not pickleable. There is also a workaround in the link.

Hope that helps!

**Comment by hlsfin** [comment]
Sorry, but this didn't work for me.
